### ISSUE TITLE
Not catching some missing return statements involving else if.

### ISSUE THREAD

**Issue by bcauchon** [davecparker/code12#282]
It seems to only happen when the code path missing the return ends in an else if after an if with a return.
The following make it through Code12's error checking:
```
	int funcMissingReturn5()
	{
		int r = 0;
		boolean boolVar1 = false;
		boolean boolVar2 = true;
		if (boolVar1)
			return r;
		else if (boolVar2)
			ct.println();
		else
			ct.println();
		// ERROR "Function is missing a return statement"
	}
	
	int funcPathMissingReturn4()
	{
		int r = 0;
		boolean boolVar1 = false;
		boolean boolVar2 = false;
		if (boolVar1)
			return r;
		else if (boolVar2)
			// ERROR "path is missing a return statement"
			ct.println();
	}
	
	int funcPathMissingReturn4WithBrackets()
	{
		int r = 0;
		boolean boolVar1 = false;
		boolean boolVar2 = false;
		if (boolVar1)
		{
			return r;
		}
		else if (boolVar2)
		{
			ct.println();
			// ERROR "path is missing a return statement"
		}
	}

**Comment by landonfox00** [comment]
Located within checkJava.lua, function checkFuncReturns

**Comment by landonfox00** [comment]
The error was held within this block of code from iLineMissingReturn
```
local s = stmt.s
if s == "return" then
	return nil   -- found required return
elseif s == "if" and stmt.elseBlock then
	-- Both sub-blocks must return a value if one does
	local iLine = iLineMissingReturn( stmt.block )
	local iLineElse = iLineMissingReturn( stmt.elseBlock )
	if not (iLine and iLineElse) then
		return iLine or iLineElse
	end
end
```
For the case of an 'else if' without an 'elseBlock,' the statement wouldn't go into the block of code where it determines if its return value is present. Instead it just jumped passed it and returned nil which is used to indicate that there was no error. This fixed the error:
```
local function iLineMissingReturn( block, isElse )
	if block then
		-- Look at the last stmt
		local stmts = block.stmts
		assert( stmts )
		local stmt = stmts[#stmts]
		if stmt == nil then
			return block.iLineEnd   -- empty block
		end
		local s = stmt.s
		if s == "return" then
			return nil   -- found required return
		elseif s == "if" and stmt.elseBlock or isElse then
			-- Both sub-blocks must return a value if one does
			local iLine = iLineMissingReturn( stmt.block, false )
			local iLineElse = iLineMissingReturn( stmt.elseBlock, true )
			if not (iLine and iLineElse) then
				return iLine or iLineElse
			end
		end
		-- This block is missing a return. Report at the } or last/only stmt
		return block.iLineEnd or stmt.iLine
	end
end
```
Its use in checkFuncReturns was also changed from
```
local iLine = iLineMissingReturn( func.block )
```
to
```
local iLine = iLineMissingReturn( func.block, false )
```

**Comment by davecparker** [comment]
Interesting, thanks for looking at this. Funny and ironic if my code to detect missing returns had a missing return in it!

But this logic is quite tricky, and I thought through it more and convinced myself it seems correct as it was. We can work through the logic together at the next meeting to clarify and see if you spot any problems.

Yet the bug remains, and with a little experimenting, I found that it only occurs if you do "else if" on a single line, not an "else" followed by a single stmt which is an "if" (although these are logically the same and produce the same internal structure).

So, I debugged further and found the actual problem was very, very subtle... When an "else if" line is parsed and built into a structure, the line number of the resulting "if" was not assigned (left nil), so then when the logic in iLineMissingReturn() correctly noticed a missing return and tried to return it, it would try to return the line number with:
```
return block.iLineEnd or stmt.iLine
```
and since there is no { block }, just a single stmt (the chained "if"), the relevant line number is stmt.iLine. But this was missing and nil in the case of "else if". Hence nil was returned when it was trying to return the line number of the code path missing the return.

So, I tried out the following fix in a totally different place. In parseProgram.lua, around line 500, where the stmt for the chained "if" is built, assign it a line number as follows (just one line of code added):
```
	-- Controlled stmts is a single stmt, which is the following if
	iTree = iTree + 1
	local stmt = { s = "if", expr = makeExpr( tree.nodes[4] ), 
				block = getControlledBlock(), 
				elseBlock = getElseBlock( ifTree ),
				entireLine = true,
				iLine = tree.iLineStart or tree.iLine }      // <== ADDED
	return { s = "block", stmts = { stmt } }
```
It seems to work, but can you try testing more cases? Especially interesting are when the last stmt in a block is an if/else with:

- both paths having a return
- only if path having a return
- only else path having a return
- neither having a return

And then "else if" used instead of else followed by if on the next line.

**Comment by landonfox00** [comment]
Good catch, I didn't consider the line numbers to be nil. After adding your fix all cases worked great.
I used three cases using this basic function, omitting certain blocks when testing specific cases:
```
int func()
{
    boolean bool = false;
    if (bool)
        return 0;
    else if (bool)
        return 1;
    else
        return 2;
}
```
The three cases are:
- if/else-if/else
- if/else-if
- if/else

In each block I used every possible sequence of placing or not placing a return statement within the block. So for the first case of if/else-if/else, I had 8 total functions. A bit overkill but it did the job haha. I can attach the code if necessary. I also tested Ben's example code and the case of a sole if statement, without an else, with a return statement within it and both of those cases were caught appropriately. The fix seems to work great.
